lib/cluster/command.js

```diff
@@ -0,0 +1,262 @@
+   * Do NOT await this method, it's meant to run indefinitely until the close()
+   * method is invoked.
+   * @returns {void}
+   */
+  async serve () {
+    this.server = new Reply();
+    await this.server.bind(`tcp://*:${this.node.config.ports.command}`);
+
+    this.state = stateEnum.RUNNING;
+
+    while (this.state === stateEnum.RUNNING) {
+      try {
+        const [type, data] = await this.server.receive();
+
+        switch (type.toString()) {
+          case topicEnum.FULLSTATE:
+            await this.sendFullState();
+            break;
+          case topicEnum.HANDSHAKE:
+            await this.handleHandshake(data);
+            break;
+          default:
+            // REP/REQ sockets expect a reply to each request made, so we have
+            // to send some kind of response on an invalid request received
+            await this.server.send([topicEnum.DISCARDED, null]);
+        }
+      }
+      catch (e) {
+        if (this.state !== stateEnum.CLOSED) {
+          throw e;
+        }
+      }
+    }
+  }
+
+  /**
+   * Closes opened sockets and marks this server as closed.
+   * @return {void}
+   */
+  dispose () {
+    this.state = stateEnum.CLOSED;
+    this.server.close();
+  }
+
+  /**
+   * Sends back the full state to the requesting node
+   * @return {void}
+   */
+  async sendFullState () {
+    const fullState = this.node.fullState.serialize();
+    fullState.activity = this.node.activity;
+    const protoEncode = this.protoroot.lookupType('FullStateResponse');
+    const buffer = protoEncode.encode(protoEncode.create(fullState)).finish();
+
+    await this.server.send([topicEnum.FULLSTATE, buffer]);
+  }
+
+  /**
+   * Handles a handshake request from a remote node
+   * @return {void}
+   */
+  async handleHandshake (data) {
+    const decoder = this.protoroot.lookupType('HandshakeRequest');
+    const { nodeId, ip, lastMessageId } = decoder.toObject(decoder.decode(data));
+
+    const added = await this.node.addNode(nodeId, ip, lastMessageId);
+
+    const encoder = this.protoroot.lookupType('HandshakeResponse');
+    const response = {
+      added,
+      lastMessageId: this.node.publisher.lastMessageId,
+    };
+
+    const buffer = encoder.encode(encoder.create(response)).finish();
+
+    await this.server.send([topicEnum.HANDSHAKE, buffer]);
+  }
+
+  /**
+   * Request the full state from ONE of the remote nodes of the cluster, AT
+   * RANDOM (necessary to make sure we distribute the full state serialization
+   * load across all nodes)
+   *
+   * @param  {Array.<IdCard>} nodes
+   * @return {Object|null} Returns the fetched full state, or null if no node answered
+   */
+  async getFullState (nodes) {
+    let idx = Math.floor(Math.random() * Math.floor(nodes.length));
+    let fullState = null;
+
+    for (let retries = 0; fullState === null && retries < nodes.length; retries++) {
+      const { id, ip } = nodes[idx];
+      const req = new Request();
+
+      req.receiveTimeout = 2000;
```

Comment:
Shouldn't we use one of the cluster config like `heartbeat` or `joinTimeout` (maybe more the heartbeat value)

Reply:
Hm, good point. 

On a cluster with very large numbers of realtime filters (so with large fullstates), we might need a dedicated timeout: `heartbeat` makes no sense here, and `joinTimeout` should be greater than the time needed to receive a single fullstate, as it's meant to limit the time for a node to join a cluster (it's used for the mutex lock).

I'll create a dedicated configuration entry.